lib-ce/emqx_modules/src/emqx_mod_rewrite.erl

```diff
@@ -75,30 +74,60 @@ description() ->
 %% Internal functions
 %%--------------------------------------------------------------------
 
-log_start(Rules) ->
-    PubRules = [{pub, Topic, Re, Dest} || {rewrite, pub, Topic, Re, Dest} <- Rules],
-    SubRules = [{sub, Topic, Re, Dest} || {rewrite, sub, Topic, Re, Dest} <- Rules],
+compile_rules(RawRules) ->
+    compile(validate_rules(RawRules)).
+
+compile({PubRules, SubRules}) ->
+    CompileRE =
+        fun({rewrite, Source, Re, Dest}) ->
+                {ok, MP} = re:compile(Re),
+                {rewrite, Source, MP, Dest}
+        end,
+    {lists:map(CompileRE, PubRules), lists:map(CompileRE, SubRules)}.
+
+validate_rules(Rules) ->
+    validate_rules(Rules, {[], []}).
+
+validate_rules([], {PubRules, SubRules}) ->
     ?LOG(info, "[Rewrite] Load: pub rules count ~p sub rules count ~p",
         [erlang:length(PubRules), erlang:length(SubRules)]),
-    log_rule(PubRules, 1),
-    log_rule(SubRules, 1).
+    log_rules(pub, PubRules),
+    log_rules(sub, SubRules),
+    {PubRules, SubRules};
+
+validate_rules([Rule = {_, Type, _, _, _} | RestRules], {PubRules, SubRules}) ->
+    case {Type, validate_rule(Rule)} of
+        {pub, {ok, NRule}} ->
+            validate_rules(RestRules, {[NRule| PubRules], SubRules});
+        {sub, {ok, NRule}} ->
+            validate_rules(RestRules, {PubRules, [NRule | SubRules]});
+        {_, {error, {Direction, T}, Reason}} ->
+            ?LOG(alert, "Topic rewrite ~p rule[~p] ~p topic ~p validate error, Reason: ~p",
+                 [assert_type(Type), rule_index(Type, PubRules, SubRules), Direction, T, Reason]),
+            {[], []}
```

Comment:
We'd better only discard the bad rule, not all of the rules if only one of them is invalid.